### PR TITLE
ISSUE-941 Tmail Plugin Reporting

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: tmail-backend/plugins/reporting/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# TMail Reports Plugin
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:1]:
```suggestion
# Mail Reports Plugin
```

File: tmail-backend/plugins/reporting/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+This plugin allows you to collect some info about the emails was processed by James mailet
+and provide webadmin routes to access these reports.
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:4]:
```suggestion
This plugin allows you to collect some info about the emails are received and sent by users
and provide webadmin routes to access these reports.
```

File: tmail-backend/plugins/reporting/README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+## How to use
+
+### Build the plugin
+
+```shell
+mvn clean install -DskipTests
+```
+
+The jar file is located in the target folder (`./target/tmail-report-plugin-jar-with-dependencies.jar`)
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:21]: No need jar with deoendency

File: tmail-backend/plugins/reporting/README.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+### Add the plugin to the TMail server
+
+- Moving the jar plugin to the TMail server (default path: `/root/extensions-jars/`)
+- Declare the plugin in the `extensions.properties` file
+
+```properties
+guice.extension.module=com.linagora.tmail.TMailReportModule
+guice.extension.startable=com.linagora.tmail.cassandra.CassandraMailReportGenerator
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:30]:
```suggestion
guice.extension.module=com.linagora.tmail.MailReportModule
guice.extension.startable=com.linagora.tmail.cassandra.CassandraMailReportGenerator
```

File: tmail-backend/plugins/reporting/README.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+### Mailet configuration
+- Customize the `mailetcontainer.xml` by adding the `TMailReceivedReportRecorder` mailet
+  eg:
+
+```xml
+
+<mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.TMailReceivedReportRecorder">
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:39]:
```suggestion
<mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.MailReportRecorder">
```

File: tmail-backend/plugins/reporting/README.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- Moving the jar plugin to the TMail server (default path: `/root/extensions-jars/`)
+- Declare the plugin in the `extensions.properties` file
+
+```properties
+guice.extension.module=com.linagora.tmail.TMailReportModule
+guice.extension.startable=com.linagora.tmail.cassandra.CassandraMailReportGenerator
+```
+
+### Mailet configuration
+- Customize the `mailetcontainer.xml` by adding the `TMailReceivedReportRecorder` mailet
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:34]:
```suggestion
- Customize the `mailetcontainer.xml` by adding the `MailReportRecorder` mailet
```

File: tmail-backend/plugins/reporting/README.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+The mailet configuration requires the following parameters:
+- `kind`: the kind of the report, which can be `SENT` or `RECEIVED`.
+
+
+### Webadmin routes configuration
+
+Declare the webadmin routes extension in the `webadmin.properties` file
+
+```properties
+extensions.routes=com.linagora.tmail.route.TMailReportsRoute
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:54]:
```suggestion
extensions.routes=com.linagora.tmail.route.MailReportsRoute
```

File: tmail-backend/plugins/reporting/README.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+### Access the webadmin routes to see the reports
+
+- URL to access the reports: `http://{webadminBaseUrl}/mu/reports/mails`
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/README.md:59]:
```suggestion
- URL to access the reports: `http://{webadminBaseUrl}/reports/mails`
```

File: tmail-backend/plugins/reporting/pom.xml [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.linagora.tmail</groupId>
+        <artifactId>tmail-plugins</artifactId>
+        <version>1.0.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>tmail-plugin-report</artifactId>
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/pom.xml:13]:
```suggestion
    <artifactId>mail-plugin-report</artifactId>
```

File: tmail-backend/plugins/reporting/pom.xml [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+                    <execution>
+                        <goals>
+                            <goal>single</goal>
+                        </goals>
+                        <phase>package</phase>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/pom.xml:156]:
```suggestion
```

Only provided dependencies: useless...

File: tmail-backend/plugins/reporting/sample_conf/mailetcontainer.xml [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                <name>X-SMIME-Status</name>
+                <onMailetException>ignore</onMailetException>
+            </mailet>
+            <mailet match="All" class="RecipientRewriteTable">
+                <errorProcessor>rrt-error</errorProcessor>
+            </mailet>
+            <mailet matcher="SenderIsLocal" class="SetMailAttribute">
+                <senderIsLocal>true</senderIsLocal>
+            </mailet>
+            <mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.TMailReceivedReportRecorder">
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/sample_conf/mailetcontainer.xml:59]:
```suggestion
            <mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.MailReportRecorder">
```

File: tmail-backend/plugins/reporting/sample_conf/mailetcontainer.xml [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                <processor>relay-denied</processor>
+            </mailet>
+        </processor>
+
+        <processor state="local-delivery" enableJmx="true">
+            <mailet match="All" class="VacationMailet"/>
+            <mailet match="All" class="Sieve"/>
+            <mailet match="All" class="AddDeliveredToHeader"/>
+            <mailet match="All" class="org.apache.james.jmap.mailet.filter.JMAPFiltering"/>
+            <mailet match="All" class="com.linagora.tmail.mailet.TMailReceivedReportRecorder">
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/sample_conf/mailetcontainer.xml:83]:
```suggestion
            <mailet match="All" class="com.linagora.tmail.mailet.MailReportRecorder">
```

File: tmail-backend/plugins/reporting/sample_conf/webadmin.properties [hunk 13]
```diff
@@ -0,0 +1,6 @@
+enabled=true
+port=8000
+host=0.0.0.0
+https.enabled=false
+
+extensions.routes=com.linagora.tmail.route.TMailReportsRoute
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/sample_conf/webadmin.properties:6]:
```suggestion
extensions.routes=com.linagora.tmail.route.MailReportsRoute
```

File: tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/TMailReportModule.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+import com.google.inject.AbstractModule;
+import com.google.inject.Scopes;
+import com.linagora.tmail.api.MailReportGenerator;
+import com.linagora.tmail.cassandra.CassandraMailReportGenerator;
+
+public class TMailReportModule extends AbstractModule {
+    @Override
+    protected void configure() {
+        bind(CassandraMailReportGenerator.class).in(Scopes.SINGLETON);
+        bind(MailReportGenerator.class).to(CassandraMailReportGenerator.class);
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/TMailReportModule.java:12]:
```suggestion
public class MailReportModule extends AbstractModule {
    @Override
    protected void configure() {
        bind(CassandraMailReportGenerator.class).in(Scopes.SINGLETON);
        bind(MailReportGenerator.class).to(CassandraMailReportGenerator.class);
```

File: tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    public static final CqlIdentifier DATE_YEAR_MONTH = CqlIdentifier.fromCql("date_year_month");
+    public static final int TTL = (int) DurationParser.parse(System.getProperty("mu.report.retention", "365d")).getSeconds();
+    public static final CassandraModule MODULE = CassandraModule.builder()
+        .table(TABLE_NAME)
+        .comment("MU sent and received mail audit trail")
+        .options(options -> options.withGcGraceSeconds(0))
+        .statement(statement -> types -> statement
+            .withPartitionKey(DATE_YEAR_MONTH, TEXT)
+            .withClusteringColumn(DATE, TIMESTAMP)
+            .withClusteringColumn(ID, TIMEUUID)
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java:54]: Please keep it STRICTLY the same

Filtering is OK and was validated with Xavier.

With this the time serie is subject to hot spots and do not scale writes BTW.

Also we do not want to migrate MU data

File: tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/mailet/TMailReceivedReportRecorder.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.james.util.ReactorUtils;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import com.linagora.tmail.api.MailReportEntry;
+import com.linagora.tmail.api.MailReportGenerator;
+
+import reactor.core.publisher.Flux;
+
+public class TMailReceivedReportRecorder extends GenericMailet {
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/mailet/TMailReceivedReportRecorder.java:19]:
```suggestion
public class MailReportRecorder extends GenericMailet {
```

File: tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/route/TMailReportsRoute.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.james.webadmin.Routes;
+import org.apache.james.webadmin.utils.JsonTransformer;
+
+import com.google.common.base.Preconditions;
+import com.linagora.tmail.api.MailReportEntry;
+import com.linagora.tmail.api.MailReportGenerator;
+
+import spark.Service;
+
+public class TMailReportsRoute implements Routes {
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/main/java/com/linagora/tmail/route/TMailReportsRoute.java:19]:
```suggestion
public class MailReportsRoute implements Routes {
```

File: tmail-backend/plugins/reporting/src/main/resources/webadmin.properties [hunk 18]
```diff
@@ -0,0 +1,5 @@
+enabled=true
+port=8000
+host=0.0.0.0
+
+extensions.routes=com.linagora.tmail.route.TMailReportsRoute
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/main/resources/webadmin.properties:5]:
```suggestion
extensions.routes=com.linagora.tmail.route.MailReportsRoute
```

File: tmail-backend/plugins/reporting/src/test/java/com/linagora/tmail/route/TMailReportsIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+import com.linagora.tmail.james.app.CassandraExtension;
+import com.linagora.tmail.james.app.DistributedJamesConfiguration;
+import com.linagora.tmail.james.app.DistributedServer;
+import com.linagora.tmail.james.app.EventBusKeysChoice;
+import com.linagora.tmail.james.app.RabbitMQExtension;
+import com.linagora.tmail.james.jmap.firebase.FirebaseModuleChooserConfiguration;
+
+import io.restassured.RestAssured;
+
+public class TMailReportsIntegrationTest {
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/test/java/com/linagora/tmail/route/TMailReportsIntegrationTest.java:57]:
```suggestion
public class MailReportsIntegrationTest {
```

File: tmail-backend/plugins/reporting/src/test/resources/extensions.properties [hunk 20]
```diff
@@ -0,0 +1,2 @@
+guice.extension.module=com.linagora.tmail.TMailReportModule
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/test/resources/extensions.properties:1]:
```suggestion
guice.extension.module=com.linagora.tmail.MailReportModule
```

File: tmail-backend/plugins/reporting/src/test/resources/mailetcontainer.xml [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+                <processor>relay-denied</processor>
+            </mailet>
+        </processor>
+
+        <processor state="local-delivery" enableJmx="true">
+            <mailet match="All" class="VacationMailet"/>
+            <mailet match="All" class="Sieve"/>
+            <mailet match="All" class="AddDeliveredToHeader"/>
+            <mailet match="All" class="org.apache.james.jmap.mailet.filter.JMAPFiltering"/>
+            <mailet match="All" class="com.linagora.tmail.mailet.TMailReceivedReportRecorder">
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/test/resources/mailetcontainer.xml:83]:
```suggestion
            <mailet match="All" class="com.linagora.tmail.mailet.MailReportRecorder">
```

File: tmail-backend/plugins/reporting/src/test/resources/mailetcontainer.xml [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+                <name>X-SMIME-Status</name>
+                <onMailetException>ignore</onMailetException>
+            </mailet>
+            <mailet match="All" class="RecipientRewriteTable">
+                <errorProcessor>rrt-error</errorProcessor>
+            </mailet>
+            <mailet matcher="SenderIsLocal" class="SetMailAttribute">
+                <senderIsLocal>true</senderIsLocal>
+            </mailet>
+            <mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.TMailReceivedReportRecorder">
```
Comment (chibenwa) [inline tmail-backend/plugins/reporting/src/test/resources/mailetcontainer.xml:59]:
```suggestion
            <mailet match="HasMailAttribute=senderIsLocal" class="com.linagora.tmail.mailet.MailReportRecorder">
```

**Review by Arsnael — COMMENTED** [review]
Read it

File: tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java [hunk 23]
```diff
@@ -93,15 +86,13 @@ private PreparedStatement prepareInsert() {
...
-            .value(DATE_YEAR_MONTH, bindMarker(DATE_YEAR_MONTH))
             .usingTtl(TTL)
             .build());
     }
 
     private PreparedStatement prepareSelectByDateBetween() {
         return session.prepare(selectFrom(TABLE_NAME)
-            .all()
-            .whereColumn(DATE_YEAR_MONTH).isEqualTo(bindMarker(DATE_YEAR_MONTH))
+            .all().allowFiltering()
```
Comment (chibenwa) [inline tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java:95]: We do not need allowFiltering without where clause

Comment (vttranlina) [reply tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java:95]: Do you mean that we will filter by java code after Cassandra fetches all, 
or do we not need to query by start and end date? (remove query by date, and remove "duration" in webadmin route)

Comment (vttranlina) [reply tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java:95]: We need allowFiltering for
```
.whereColumn(DATE).isGreaterThanOrEqualTo(bindMarker(DATE_START))
            .whereColumn(DATE).isLessThanOrEqualTo(bindMarker(DATE_END))
```

Comment (chibenwa) [reply tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/cassandra/CassandraMailReportGenerator.java:95]: > Do you mean that we will filter by java code after Cassandra fetches all

Yes.

Easier: it avoids timeouts

File: tmail-backend/plugins/mail-reporting/pom.xml [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+        <version>1.0.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>mail-plugin-report</artifactId>
+    <name>Mail Report Plugin</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (HoussemNasri) [inline tmail-backend/plugins/mail-reporting/pom.xml:18]:
```suggestion
            <groupId>${james.groupId}</groupId>
```

File: tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/mailet/MailReportRecorder.java [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        String subject = Optional.ofNullable(mail.getMessage().getSubject()).orElse("<no subject>");
+
+        Flux.fromIterable(mail.getRecipients())
+            .map(recipient -> new MailReportEntry(kind,
+                subject,
+                mail.getMaybeSender(),
+                recipient,
+                instant))
+            .flatMap(mailReportGenerator::append, ReactorUtils.DEFAULT_CONCURRENCY)
+            .then()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply tmail-backend/plugins/mail-reporting/src/main/java/com/linagora/tmail/mailet/MailReportRecorder.java:49]: IMO I prefer this syntax to the `.blockLast` alternative...

**Review by quantranhong1999 — APPROVED** [review]
Good job

**Review by chibenwa — APPROVED** [review]

**Pull request merged**